subsys/bluetooth/controller/ll_sw/ull_adv_iso.c

```diff
@@ -10,28 +10,396 @@
 		      uint8_t packing, uint8_t framing, uint8_t encryption,
-		      uint8_t *bcode)
+		      uint8_t *bcode, void **rx)
 {
-	/* TODO: Implement */
-	ARG_UNUSED(big_handle);
-	ARG_UNUSED(adv_handle);
-	ARG_UNUSED(num_bis);
-	ARG_UNUSED(sdu_interval);
-	ARG_UNUSED(max_sdu);
-	ARG_UNUSED(max_latency);
-	ARG_UNUSED(rtn);
-	ARG_UNUSED(phy);
-	ARG_UNUSED(packing);
-	ARG_UNUSED(framing);
-	ARG_UNUSED(encryption);
-	ARG_UNUSED(bcode[16]);
+	struct ll_adv_iso *adv_iso;
+	struct ll_adv_set *adv;
+	struct node_rx_pdu *node_rx;
+	struct pdu_bis *pdu_bis;
+	struct node_tx *tx;
 
-	return BT_HCI_ERR_CMD_DISALLOWED;
+	adv_iso = ull_adv_iso_get(big_handle);
+
+	if (!adv_iso) {
+		return BT_HCI_ERR_CMD_DISALLOWED;
+	}
+
+	adv = ull_adv_is_created_get(adv_handle);
+
+	if (!adv) {
+		return BT_HCI_ERR_UNKNOWN_ADV_IDENTIFIER;
+	} else if (!adv->lll.sync) {
+		return BT_HCI_ERR_UNKNOWN_ADV_IDENTIFIER;
+	} else if (adv->lll.sync->adv_iso) {
+		return BT_HCI_ERR_UNKNOWN_ADV_IDENTIFIER;
+	}
+
+	if (IS_ENABLED(CONFIG_BT_CTLR_PARAM_CHECK)) {
+		if (num_bis == 0 || num_bis > 0x1F) {
+			return BT_HCI_ERR_INVALID_PARAM;
+		}
+
+		if (sdu_interval < 0x000100 || sdu_interval > 0x0FFFFF) {
+			return BT_HCI_ERR_INVALID_PARAM;
+		}
+
+		if (max_sdu < 0x0001 || max_sdu > 0x0FFF) {
+			return BT_HCI_ERR_INVALID_PARAM;
+		}
+
+		if (max_latency > 0x0FA0) {
+			return BT_HCI_ERR_INVALID_PARAM;
+		}
+
+		if (rtn > 0x0F) {
+			return BT_HCI_ERR_INVALID_PARAM;
+		}
+
+		if (phy > 3) {
+			return BT_HCI_ERR_INVALID_PARAM;
+		}
+
+		if (packing > 1) {
+			return BT_HCI_ERR_INVALID_PARAM;
+		}
+
+		if (framing > 1) {
+			return BT_HCI_ERR_INVALID_PARAM;
+		}
+
+		if (encryption > 1) {
+			return BT_HCI_ERR_INVALID_PARAM;
+		}
+	}
+
+	/* TODO: Allow more than 1 BIS in a BIG */
+	if (num_bis != 1) {
+		return BT_HCI_ERR_MEM_CAPACITY_EXCEEDED;
+	}
+	/* TODO: For now we can just use the unique BIG handle as the BIS
+	 * handle until we support multiple BIS
+	 */
+	adv_iso->bis_handle = big_handle;
+
+	adv_iso->num_bis = num_bis;
+	adv_iso->sdu_interval = sdu_interval;
+	adv_iso->max_sdu = max_sdu;
+	adv_iso->max_latency = max_latency;
+	adv_iso->rtn = rtn;
+	adv_iso->phy = phy;
+	adv_iso->packing = packing;
+	adv_iso->framing = framing;
+	adv_iso->encryption = encryption;
+	memcpy(adv_iso->bcode, bcode, sizeof(adv_iso->bcode));
+
+	/* TODO: Add ACAD to AUX_SYNC_IND */
+
+	/* TODO: start sending BIS empty data packet for each BIS */
+	ull_adv_iso_start(adv_iso, 0 /* TODO: Calc ticks_anchor */);
+
+	tx = ll_iso_tx_mem_acquire();
+	if (!tx) {
+		BT_ERR("Tx Buffer Overflow");
+		return BT_HCI_ERR_MEM_CAPACITY_EXCEEDED;
+	}
+
+	adv_iso->is_created = true;
+	pdu_bis = (void *)tx->pdu;
+
+	/* Empty BIS packet */
+	pdu_bis->length = 0;
+	pdu_bis->ll_id = PDU_BIS_LLID_COMPLETE_END;
+	pdu_bis->cssn = 0;
+	pdu_bis->cstf = 0;
+
+	if (ll_iso_tx_mem_enqueue(adv_iso->bis_handle, tx)) {
+		BT_ERR("Invalid Tx Enqueue");
+		ll_iso_tx_mem_release(tx);
+		return BT_HCI_ERR_CMD_DISALLOWED;
+	}
+
```

Comment:
I guess this way of allocating is okay for this very first iteration. However, I think we should discuss how ISO buffers should be handled in the final BIS/CIS implementation to make sure we do not go in a direction now that will make it difficult to do things right in the end.

Reply:
I agree. This way is heavily inspired of how it's done for connection PDUs. Synchronizing how it's done for connection and ISO packets seems alright to me.